specification/src/main/asciidoc/jakarta-concurrency.adoc

```diff
@@ -2302,3 +2302,159 @@ ends.
 
 See section 3.1.8.2.1 for an example of using a `UserTransaction` within a
 task.
+
+== Asynchronous Methods
+
+The `jakarta.enterprise.concurrent.Async` annotation annotates a
+CDI managed bean method to run asynchronously or annotates a
+CDI managed bean class with methods to run asynchronously.
+
+Each asynchronous method execution corresponds to a managed
+`java.util.concurrent.CompletableFuture` instance that is backed by a
+`jakarta.enterprise.concurrent.ManagedExecutorService` as its default
+asynchronous execution facility. Its dependent stages
+(and all dependent stages that are created from those, and so on)
+continue to be backed by the managed executor service,
+which also manages the propagation of context to completion stage actions.
+Application code, including from within the asynchronous method, can query
+the status of the completable future instance and can choose to complete
+it at any time and by any means.
+
+==== Application Component Provider’s Responsibilities
+
+Application Component Providers (application developers) (EE2.11.2)
+specify the `jakarta.enterprise.concurrent.Async` annotation on
+CDI managed bean methods with return type of
+`java.util.concurrent.CompletableFuture`,
+`java.util.concurrent.CompletionStage`, or `void` to designate them for
+asynchronous execution. Alternatively, the
+`jakarta.enterprise.concurrent.Async` annotation can be specified on a
+CDI managed bean class to designate all methods with return type of
+`java.util.concurrent.CompletableFuture` or
+`java.util.concurrent.CompletionStage` for asynchronous execution.
```

Comment:
Should it also turn all void methods asynch?

Reply:
I was a bit hesitant about even allowing the asynchronous methods annotation at class level at all, except that it is useful there for the CDI extension to annotate a class as the interceptor.
Once we have it at class level, then there's a risk that the user will end up with methods inadvertently becoming asynchronous that they didn't expect to be.  For example, method `String getName()` isn't asynchronous (due to String return type), and if a user adds 'void setName(String name)` it likely won't even cross their mind that the method would be asynchronous, which could end up causing intermittent bugs due to the unexpected parallelism.  The opposite could happen too, but that seemed both less likely and less severe.
I wasn't sure what the best approach is here, so I'm glad to have others thinking through it as well.
Another option is to have the spec state that applications are not to specify the asynchronous methods annotation at class level (this is only for the CDI extension to do in defining an interceptor) and that applications must only define it at the method level. This could be easily enforced by the interceptor and remove the ambiguity over which methods do/don't become asynchronous, at the trade-off of possibly being confusing because the annotation would still have `ElementType.TYPE` in its `@Target`.
What is your preference?

Reply:
I think not allowing a class level annotation is probably the safest solution from the perspective of the developer using the api in that side effects are reduced. Do you think we can detect and specify a deployment exception should be thrown if application code marks it at the class level?

I know people complain about verbosity but personally I think explicitly marking methods with asynchronous would be best practice anyway and easier for anybody having to maintain the application code.

Reply:
> I know people complain about verbosity but personally I think explicitly marking methods with asynchronous would be best practice anyway and easier for anybody having to maintain the application code.

Since we both agree that limiting to method level is the best practice, let's go with that approach.  It will simplify the API and make application code more readable and explicit about the behavior it is getting.

> I think not allowing a class level annotation is probably the safest solution from the perspective of the developer using the api in that side effects are reduced. Do you think we can detect and specify a deployment exception should be thrown if application code marks it at the class level?

I know for sure that an interceptor can at least detect it at runtime and raise an error (such as UnsupportedOperationException) upon the attempt to invoke a method on the invalidly annotated class.  I'll need to look into whether there is a way to detect it upon deployment to raise a deployment exception then.